livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -208,7 +208,22 @@ private SessionInfo getSessionInfo(int sessionId) throws LivyException {
 
   public InterpreterResult interpret(String code, boolean displayAppInfo)
       throws LivyException {
-    StatementInfo stmtInfo = executeStatement(new ExecuteRequest(code));
+    StatementInfo stmtInfo = null;
+    try {
+      stmtInfo = executeStatement(new ExecuteRequest(code));
+    } catch (SessionNotFoundException e) {
+      LOGGER.warn("Livy session {} is expired, new session will be created.", sessionInfo.id);
+      this.sessionExpired = true;
```

Comment:
instead of saving this as a state, would be more self-contain to pass this as a flag to `appendSessionExpire` below?